ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -315,6 +316,7 @@ class WebSocketService(
         .filter(_.nonEmpty)
```

Comment:
This occurring earlier in the stream means that empty (or emptied by means of filtering) steps will not cause offset to "tick up". You might have to `-\/` on the heartbeat before this stage in the flow.

Reply:
@S11001001 @hurryabit  is that really important to "tick up" empty events? You will get the heartbeat in any case if there is no activity matching the query. We either filter empty events out completely or we need to add some logic for conflating consecutive empty events, which would rather complicate the flow.

If you only care about the "live" marker, it can be sent explicitly.

Reply:
It is important, and a prime motivation for this feature.

Consider the sequence of event blocks, before we filter out the empty ones. Each has a unique offset, representing when it happened, and therefore when the stream will pick up if you resume it.

A block becomes empty in the stream for only one reason: we filtered out all the events in it. This feature is important to preserve so you don't get spammed with empty event blocks should you choose a particularly strict query (e.g. all key queries are quite strict).

Suppose that you get a block, and then four hours pass. Suppose that there are empty event blocks in that time. Under the current code, the heartbeat will continue to emit the offset from the last non-empty event block. What we want is for it to emit the offset from the last _empty or non-empty_ event block, so that should you close the stream and resume, you won't necessarily be four hours in the past.

I don't think this has to terribly affect the current flow; your present logic can work on `StepAndErrors`, possibly making `StepAndErrors` a full ADT (or not, do whatever works structurally), rather than the output of `render`, and an "empty" `StepAndErrors` that should be emitted anyway because it is a heartbeat can simply be `nonEmpty`.

Reply:
See: #4955